tools/jnimarshalmethod-gen/TypeMover.cs

```diff
@@ -143,6 +143,34 @@ static TypeDefinition Resolve (TypeReference type)
 			return resolved;
 		}
 
+		static TypeReference GetUpdatedArray (TypeReference type, TypeReference newReference)
+		{
+			var t = type;
+			var stack = new Stack<ArrayType> ();
+
+			do {
+				var ta = t as ArrayType;
+
+				stack.Push (ta);
+				t = ta.ElementType;
+			} while (t.IsArray);
+
+			while (stack.Count > 0) {
+				var ta = stack.Pop ();
+				var na = new ArrayType (newReference, ta.Rank);
+
+				if (ta.Rank > 1) {
```

Comment:
> Or is GetUpdatedArray() invoked for members that aren't marshal methods?

It is now probably used only by marshal methods. It is not a good place to check for multidimensional arrays though as it can be used for other types as well. The `__RegisterNativeMembers` code might change to use arrays for example.

The `jnimarshalmethod-gen` tool looks like a good place to reject `[Export]` members with *md* arrays though, so I opened https://github.com/xamarin/java.interop/issues/353 for that.